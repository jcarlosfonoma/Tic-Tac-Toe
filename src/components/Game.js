import Board from "./Board";
import styled from "styled-components";

const GameInfo = styled.div`
  margin-left: 20px;
  background-color: #3d4250;
  border-radius: 12px;
  color: white;
`;

const Status = styled.div`
  margin-bottom: 10px;
  margin-top: 10px;
  text-align: center;
  font-weight: bold;
`;

const PLayList = styled.ol`
  list-style: none;
  padding: 0;
`;

const Game = (props) => {
  return (
    <div className="game">
      <div className="game-board">
        <Board data-testid="board-btn" squares={props.squares} onClick={(i) => props.onClick(i)} />
      </div>
      <GameInfo>
        <Status data-testid="status-div">{props.status}</Status>
        <PLayList>{props.moves}</PLayList>
      </GameInfo>
    </div>
  );
};

export default Game;
