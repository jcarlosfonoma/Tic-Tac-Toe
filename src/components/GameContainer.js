import React, { useState, useEffect, useContext } from "react";
import Game from "./Game";
import MyContext from "../store/matches-record-context";
import styled from "styled-components";

const GameContainer = () => {
  const [history, setHistory] = useState([{ squares: Array(9).fill(null) }]);
  const [stepNumber, setStepNumber] = useState(0);

  const matchesResultCtx = useContext(MyContext);

  const matchResultHandler = (winner = null) => {
    matchesResultCtx.winner(winner);
  };

  const handleClick = (i) => {
    const auxHistory = history.slice(0, stepNumber + 1);
    const current = auxHistory[auxHistory.length - 1];
    const squares = current.squares.slice();

    if (winner || squares[i]) {
      return;
    }

    squares[i] = matchesResultCtx.xIsNext ? "X" : "O";

    setHistory(auxHistory.concat([{ squares }]));
    setStepNumber(auxHistory.length);

    matchesResultCtx.nextHandler();
  };

  const jumpTo = (move) => {
    setStepNumber(move);
    matchesResultCtx.nextHandler(move % 2 === 0);
  };

  const calculateWinner = (squares) => {
    const lines = [
      [0, 1, 2],
      [3, 4, 5],
      [6, 7, 8],
      [0, 3, 6],
      [1, 4, 7],
      [2, 5, 8],
      [0, 4, 8],
      [2, 4, 6],
    ];

    for (let i = 0; i < lines.length; i++) {
      const [a, b, c] = lines[i];
      if (
        squares[a] &&
        squares[a] === squares[b] &&
        squares[a] === squares[c]
      ) {
        return squares[a];
      }
    }
    return null;
  };

  let current = history[stepNumber];
  let winner = calculateWinner(current.squares);

  useEffect(() => {
    if (winner || stepNumber === 9) {
      matchResultHandler(winner);
    }
  }, [winner, stepNumber]);

  let status = winner
    ? `Winner ${winner}`
    : `Next player ${matchesResultCtx.xIsNext ? "X" : "O"}`;

  const moves = history.map((step, move) => {
    const description = move ? `Go to move # ${move}` : "Go to game start";
    return (
      <MoveItem key={move}>
        <MoveBtn onClick={() => jumpTo(move)}>{description}</MoveBtn>
      </MoveItem>
    );
  });

  return (
    <Game
      squares={current.squares}
      onClick={handleClick}
      status={status}
      moves={moves}
    />
  );
};

const MoveItem = styled.li`
  margin: 10px;
  padding: 0;
`;

const MoveBtn = styled.button`
  padding: 4px;
  width: 100%;
  background-color: #3d4250;
  color: white;
  height: 30px;
  border-radius: 5px;
`;

export default GameContainer;
