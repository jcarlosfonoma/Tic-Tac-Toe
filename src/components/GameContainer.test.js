import { cleanup, render, fireEvent } from "@testing-library/react";
import GameContainer from "./GameContainer";
import Game from "./Game";

describe("GameContainer component", () => {
  beforeEach(cleanup);

  it("GameContainer should take a snapshot", () => {
    const { asFragment } = render(<GameContainer />);
    expect(asFragment(<GameContainer />)).toMatchSnapshot();
  });

  //   it("Checks if initial status is equal to Go to game start", () => {
  //     const { getByTestId } = render(<GameContainer />);
  //     expect(getByTestId("status-div")).toHaveTextContent("Go to game start");
  //   });

  it("Checks if onClick event is called", () => {
    const onClick = jest.fn();

    const squares = Array(9).fill(null);

    const { getAllByTestId } = render(<Game squares={squares} onClick={onClick}/>);
    const square = getAllByTestId('square-btn')[0];

    fireEvent.click(square);
    expect(onClick).toHaveBeenCalled();
  });
});
