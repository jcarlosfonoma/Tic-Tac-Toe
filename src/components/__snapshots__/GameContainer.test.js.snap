// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameContainer component GameContainer should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="game"
  >
    <div
      class="game-board"
    >
      <div>
        <div
          class="board-row"
        >
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
        </div>
        <div
          class="board-row"
        >
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
        </div>
        <div
          class="board-row"
        >
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
          <button
            class="sc-bdvvtL fPtRnW"
            data-testid="square-btn"
          />
        </div>
      </div>
    </div>
    <div
      class="sc-gsDKAQ gKFSjR"
    >
      <div
        class="sc-dkPtRN hDBrjf"
        data-testid="status-div"
      >
        Next player X
      </div>
      <ol
        class="sc-hKwDye fqVKqO"
      >
        <li
          class="sc-eCImPb jiTVab"
        >
          <button
            class="sc-jRQBWg bShOYO"
          >
            Go to game start
          </button>
        </li>
      </ol>
    </div>
  </div>
</DocumentFragment>
`;
